detect_test.go

```diff
@@ -93,33 +67,95 @@ func testDetect(t *testing.T, context spec.G, it spec.S) {
-	}, spec.Sequential())
 
-	context("error handling", func() {
-		it("handles an error from the pyproject.toml parser", func() {
-			expectedErr := errors.New("hi")
-			parsePythonVersion.ParsePythonVersionCall.Returns.Error = expectedErr
+		context("when the BP_POETRY_VERSION is set", func() {
+			it.Before(func() {
+				Expect(os.Setenv("BP_POETRY_VERSION", "some-version")).To(Succeed())
+			})
 
-			result, err := detect(packit.DetectContext{
-				WorkingDir: "/working-dir",
+			it.After(func() {
+				Expect(os.Unsetenv("BP_POETRY_VERSION")).To(Succeed())
+			})
+
+			it("returns a plan that requires that version of poetry", func() {
+				parsePythonVersion.ParsePythonVersionCall.Returns.String = "9.8.7"
+
+				result, err := detect(packit.DetectContext{
+					WorkingDir: workingDir,
+				})
+				Expect(err).NotTo(HaveOccurred())
+				Expect(parsePythonVersion.ParsePythonVersionCall.Receives.String).To(Equal(filepath.Join(workingDir, "pyproject.toml")))
+				Expect(result).To(Equal(packit.DetectResult{
+					Plan: packit.BuildPlan{
+						Provides: []packit.BuildPlanProvision{
+							{Name: "poetry"},
+						},
+						Requires: []packit.BuildPlanRequirement{
+							{
+								Name: poetry.Pip,
+								Metadata: poetry.BuildPlanMetadata{
+									Build: true,
+								},
+							},
+							{
+								Name: poetry.CPython,
+								Metadata: poetry.BuildPlanMetadata{
+									Build:         true,
+									Version:       "9.8.7",
+									VersionSource: "pyproject.toml",
+								},
+							},
+							{
+								Name: "poetry",
+								Metadata: poetry.BuildPlanMetadata{
+									VersionSource: "BP_POETRY_VERSION",
+									Version:       "some-version",
+								},
+							},
+						},
+					},
+				}))
+			})
+		})
+
+		context("error handling", func() {
+			// the python dependency is mandatory
+			// https://python-poetry.org/docs/pyproject/#dependencies-and-dev-dependencies
+			it("fails detection when no python version found", func() {
+				parsePythonVersion.ParsePythonVersionCall.Returns.String = ""
+
+				_, err := detect(packit.DetectContext{
+					WorkingDir: workingDir,
+				})
+				Expect(err).To(MatchError(packit.Fail.WithMessage("pyproject.toml must include [tool.poetry.dependencies.python], see https://python-poetry.org/docs/pyproject/#dependencies-and-dev-dependencies")))
+			})
+
+			it("handles an error from the pyproject.toml parser", func() {
+				expectedErr := errors.New("hi")
+				parsePythonVersion.ParsePythonVersionCall.Returns.Error = expectedErr
+
+				_, err := detect(packit.DetectContext{
+					WorkingDir: workingDir,
+				})
+				Expect(err).To(Equal(expectedErr))
+			})
+		})
+	})
+
+	context("when pyproject.toml is not present", func() {
+		it("fails detection", func() {
+			_, err := detect(packit.DetectContext{
+				WorkingDir: workingDir,
 			})
-			Expect(result).To(Equal(packit.DetectResult{}))
-			Expect(err).To(Equal(expectedErr))
+			Expect(err).To(MatchError("pyproject.toml is not present"))
```

Comment:
Similar to the previous comment, I would move this to the happy-path section. I'd also validate the specific error is `packit.Fail` rather than the more general assertion we see here.

Reply:
@robdimsdale do you have an example of how to validate that specific error? I tried a few permutations without success.

Reply:
This context should already be at the top level. The context for `error handling` is in another context.

Reply:
I was thinking of doing something similar to what we see above e.g.:

```
Expect(err).To(MatchError(packit.Fail.WithMessage("pyproject.toml file is not present")))
```

Reply:
As discussed, we can modify this at a later date.